### DISCUSSION TITLE
How to prevent build reading env variables?

### DISCUSSION THREAD

**Discussion by OperationalFallacy** [vercel/next.js#37723] [Help]
Hi,

We're developing a site and wanted to make sure it doesn't pull any data from headless CMS during the build.
The most assurance would be to have access to env variables disabled.
Is there a config option to do it? 

Thanks!

**Comment by icyJoseph** [comment]
If you don't have `getStaticProps`, then it won't pull data during build. 

Could you explain a little more on what is it you are trying to do and why?

If there's no env vars at build time, then undefined will be put in place of env vars. Perhaps runtime configuration, https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration, might be an option, but I doubt it does what you want it to do.

**Comment by OperationalFallacy** [comment]
We're not super proficient with Next yet, so this is just a safety measure if somebody accidentally adds getStaticProps.

**Reply by icyJoseph** [reply]
Safety in which sense? Are you afraid of leaking secrets to the client? Or just avoid getting any data from the CMS, at build time because it will become stale rather quick? 

I don't know of any global config to disable getStaticProps though.

This tool, https://next-code-elimination.vercel.app/, shows what code will be bundled to the client, and what'll only exist on the server, or build time .

**Reply by OperationalFallacy** [reply]
We don't want to pull cms data into the build.
Nextjs bundles data blobs with the build when getStaticProps is present.
Maybe I want a strange thing :)
Thanks for the app, its useful

**Reply by icyJoseph** [reply]
No, I think you have a point. GetStaticProps resolves to JSON, which is included on the page, and sent from the server to the client when SPA transition occurs to a page that used GetStaticProps.

However, I don't think there's a good automated way to prevent this. 

When the app is built, a summary is printed. For example this from the test bench mark I use to solve questions posted by people here:

```
Page                                       Size     First Load JS
┌ λ /                                      337 B          77.7 kB
├   /_app                                  0 B            74.9 kB
├ ƒ /_middleware                           0 B            74.9 kB
├ ○ /404                                   193 B          75.1 kB
├ λ /api/hello                             0 B            74.9 kB
├ λ /api/revalidate                        0 B            74.9 kB
├ ○ /blog                                  297 B          77.7 kB
└ ● /item (ISR: 10 Seconds)                938 B          75.9 kB
+ First Load JS shared by all              74.9 kB
  ├ chunks/framework-a87821de553db91d.js   45 kB
  ├ chunks/main-fc7d2f0e2098927e.js        28.7 kB
  ├ chunks/pages/_app-43ac001d99fedbd3.js  493 B
  ├ chunks/webpack-69bfa6990bb9e155.js     769 B
  └ css/27d177a30947857b.css               194 B

ƒ  (Middleware)  intercepts requests (uses _middleware)
λ  (Server)      server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)      automatically rendered as static HTML (uses no initial props)
●  (SSG)         automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)         incremental static regeneration (uses revalidate in getStaticProps)
```

I guess, you could pipe that result into a bash script or something and if you find `getStaticProps` in the text, you can exit with `1`, and prevent deploys?

### Edit

```fish
➜ if yarn build | rg getStaticProps
      echo App is using GSP
  else
      echo GSP was not used # yarn deploy perhaps?
  end
```

For example this FISH shell script, if the build output, when piped to `ripgrep` yields a match for `getStaticProps`, then you could echo and do nothing further, otherwise, deploy, or use a further exit code, idk, depends on your bash/shell scripts skill, and how would you deploy this application.

This would be a bit brittle though.

**Reply by OperationalFallacy** [reply]
Haha yeah, these bash scripts can turn a very nice pipeline-less cicd experience for devs into a jenkins style nightmare :)
